### PR TITLE
Alien 1910 - Resolve dependency conflicts when uploading archives

### PR DESCRIPTION
**Pull request by afraisse**
Reverted Igor's commit for ALIEN-1867.
Reimplemented his processImports method to deal with dependency conflicts using the following rules:
- direct dependency conflicts with transitive dependency => Use direct dependency version
- direct dependencies conflict with one another  => Use the latest version
- transitive dependency conflicts with one another => Use the latest version

### CODE REVIEW

File: alien4cloud-tosca/src/main/java/alien4cloud/tosca/parser/postprocess/ArchiveRootPostProcessor.java [hunk 1]
```diff
@@ -88,30 +87,95 @@ private void doProcess(ArchiveRoot archiveRoot) {
...
-                            .findFirst().ifPresent(currentDependency -> {
-                                ParsingContextExecution.getParsingErrors()
-                                        .add(new ParsingError(ErrorCode.DEPENDENCY_VERSION_CONFLICT, csar.getId(), node.getStartMark(),
-                                                AlienUtils.prefixWith(":", transitiveDependency.getVersion(), transitiveDependency.getName()),
-                                                node.getEndMark(), AlienUtils.prefixWith(":", currentDependency.getVersion(), currentDependency.getName())));
-                            });
+        /* Three types of conflicts :
+          - A transitive dep has a different version than a direct dependency => Force transitive to direct version
+          - Transitive dependencies with the same name and different version are used => Use latest
+          - Direct dependencies with the same name and different version are used => Error or use latest ?
```
Comment (lucboutier) [inline alien4cloud-tosca/src/main/java/alien4cloud/tosca/parser/postprocess/ArchiveRootPostProcessor.java]: This is the 1. in the following code, should it be marked as 1. ?

File: alien4cloud-tosca/src/main/java/alien4cloud/tosca/parser/impl/ErrorCode.java [hunk 2]
```diff
@@ -72,6 +72,7 @@
     INVALID_NODE_TEMPLATE_NAME,
     /** A node type a defined in other Csar **/
     TOSCA_TYPE_ALREADY_EXISTS_IN_OTHER_CSAR,
-    /** version conflict between a transitive and a direct dependency **/
```
Comment (lucboutier) [inline alien4cloud-tosca/src/main/java/alien4cloud/tosca/parser/impl/ErrorCode.java]: Not intentional to remove this ? DEPENDENCY_VERSION_CONFLICT is not commented/explained anymore.

Comment (afraisse) [inline alien4cloud-tosca/src/main/java/alien4cloud/tosca/parser/impl/ErrorCode.java]: Not intentional indeed.

File: alien4cloud-tosca/src/main/java/alien4cloud/tosca/parser/postprocess/ArchiveRootPostProcessor.java [hunk 3]
```diff
@@ -88,30 +87,95 @@ private void doProcess(ArchiveRoot archiveRoot) {
...
+                .filter(c -> c != null)
+                .reduce(Sets::union)
+                .orElse(Collections.emptySet())
+        );
+
+        // 2. Resolve all transitive vs. direct dependencies conflicts with foreach side-effects using the direct dependency's version
+        transitiveDependencies.forEach(transitiveDependency -> dependencies.stream()
+                .filter(directDep -> Objects.equals(directDep.getName(), transitiveDependency.getName())
+                        && !Objects.equals(directDep.getVersion(), transitiveDependency.getVersion()))
+                .findFirst() // Has we resolved direct dependencies conflicts earlier, there can only be one direct dependency that conflicts
```
Comment (lucboutier) [inline alien4cloud-tosca/src/main/java/alien4cloud/tosca/parser/postprocess/ArchiveRootPostProcessor.java]: Has => as :)

**Pull request merged**